### PR TITLE
[stackalloc-init] Use initblk whenever possible regardless of type size

### PR DESCRIPTION
**Pull request by alrz**

### CODE REVIEW

**Comment by alrz** [comment]
This is partially blocked on https://github.com/dotnet/roslyn/issues/23951

**Comment by VSadov** [comment]
Right. 
The ETA on [https://github.com/dotnet/corefx/issues/26948](https://github.com/dotnet/corefx/issues/26948)  is not clear. Likely not in the immediate release. FX changes and especially new APIs that may have platform-specific implementations need longer lead time.

For the [https://github.com/dotnet/roslyn/issues/23951](https://github.com/dotnet/roslyn/issues/23951) - it would be actionable when these two features meet. - one or another would need to be in master. It is optional though. We can do that when both are in master.

---
In reply to: [366472308](https://github.com/dotnet/roslyn/pull/24915#issuecomment-366472308) [](ancestors = 366472308)

**Comment by VSadov** [comment]
@dotnet-bot test windows_release_unit32_prtest  please

**Comment by VSadov** [comment]
@dotnet-bot all tests have passed. But then the build failed. Possibly some network issue...

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler - can we get a second review on this?

File: src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs [hunk 1]
```diff
@@ -30,11 +30,22 @@ private void EmitStackAllocInitializers(TypeSymbol type, BoundArrayInitializatio
             else
             {
                 ImmutableArray<byte> data = this.GetRawData(initExprs);
-                _builder.EmitStackAllocBlockInitializer(data, inits.Syntax, _diagnostics);
+                if (data.All(datum => datum == data[0]))
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs]: How are we assured that data is non empty here?

File: src/Compilers/CSharp/Test/Emit/CodeGen/StackAllocInitializerTests.cs [hunk 2]
```diff
@@ -58,35 +58,35 @@ public void TestIdenticalBytes()
 
 static unsafe class C
 {
-    static void Print(byte* p)
+    static void Print(uint* p)
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit/CodeGen/StackAllocInitializerTests.cs]: Why did we change this test vs. adding a new one?

File: src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs (hunk 1 above)
Comment (alrz) [reply src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs]: The codepath is unreachable for initializers for less than two elements.

File: src/Compilers/CSharp/Test/Emit/CodeGen/StackAllocInitializerTests.cs [hunk 3]
```diff
@@ -51,7 +51,40 @@ .maxstack  1
...
+            CompileAndVerify(text,
+                parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7_3),
+                options: TestOptions.UnsafeReleaseExe,
+                verify: Verification.Fails).VerifyIL("C.Main",
+@"{
+  // Code size        8 (0x8)
+  .maxstack  1
+  IL_0000:  ldc.i4.0
+  IL_0001:  conv.u
+  IL_0002:  call       ""void C.Use(int*)""
```
Comment (alrz) [inline src/Compilers/CSharp/Test/Emit/CodeGen/StackAllocInitializerTests.cs]: this is identical to what `stackalloc int[0]` produces.

**Comment by VSadov** [comment]
Thanks!!

**Pull request merged**